libraries/shared/factories/event/repository.go

```diff
@@ -16,9 +16,158 @@
+	ColumnValues   ColumnValues // Associated values for columns, restricted to []byte, bool, float64, int64, string, time.Time
+}
+
+// ModelToQuery stores memoised insertion queries to minimise computation
+var ModelToQuery = map[string]string{}
+
+// GetMemoizedQuery gets/creates a DB insertion query for the model
+func GetMemoizedQuery(model InsertionModel) string {
+	// The schema and table name uniquely determines the insertion query, use that for memoization
+	queryKey := string(model.SchemaName) + string(model.TableName)
+	query, queryMemoized := ModelToQuery[queryKey]
+	if !queryMemoized {
+		query = GenerateInsertionQuery(model)
+		ModelToQuery[queryKey] = query
+	}
+	return query
+}
+
+// GenerateInsertionQuery creates an SQL insertion query from an insertion model.
+// Should be called through GetMemoizedQuery, so the query is not generated on each call to Create.
+func GenerateInsertionQuery(model InsertionModel) string {
+	var valuePlaceholders []string
+	var updateOnConflict []string
+	for i := 0; i < len(model.OrderedColumns); i++ {
+		valuePlaceholder := fmt.Sprintf("$%d", 1+i)
+		valuePlaceholders = append(valuePlaceholders, valuePlaceholder)
+		updateOnConflict = append(updateOnConflict,
+			fmt.Sprintf("%s = %s", model.OrderedColumns[i], valuePlaceholder))
+	}
+
+	baseQuery := `INSERT INTO %v.%v (%v) VALUES(%v)
+		ON CONFLICT (header_id, log_id) DO UPDATE SET %v;`
+
+	return fmt.Sprintf(baseQuery,
+		model.SchemaName,
+		model.TableName,
+		joinOrderedColumns(model.OrderedColumns),
+		strings.Join(valuePlaceholders, ", "),
+		strings.Join(updateOnConflict, ", "))
+}
+
+/*
+Create generates an insertion query and persists to the DB, given a slice of InsertionModels.
+ColumnValues are restricted to []byte, bool, float64, int64, string, time.Time.
+
+testModel = shared.InsertionModel{
+	SchemaName:     "public"
+	TableName:      "testEvent",
+	OrderedColumns: []string{"header_id", "log_id", "variable1"},
+	ColumnValues: ColumnValues{
+		"header_id": 303
+		"log_id":   "808",
+		"variable1": "value1",
+	},
+}
+*/
+func Create(models []InsertionModel, db *postgres.DB) error {
+	if len(models) == 0 {
+		return fmt.Errorf("repository got empty model slice")
+	}
+
+	tx, dbErr := db.Beginx()
+	if dbErr != nil {
+		return dbErr
+	}
+
+	for _, model := range models {
+		// Maps can't be iterated over in a reliable manner, so we rely on OrderedColumns to define the order to insert
+		// tx.Exec is variadically typed in the args, so if we wrap in []interface{} we can apply them all automatically
+		var args []interface{}
+		for _, col := range model.OrderedColumns {
+			value := model.ColumnValues[col]
+			// Check whether or not PG can accept the type of value in the model
+			okPgValue := driver.IsValue(value)
+			if !okPgValue {
+				logrus.WithField("model", model).Errorf("PG cannot handle value of this type: %T", value)
+				return ErrUnsupportedValue
```

Comment:
I dig the log above, but am also wondering if we might want to build this error as a function that takes the model/value and includes that data there too - just thinking it could make the problem more obvious more quickly if it's in the error itself rather than just the logs